keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -0,0 +1,1026 @@
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+The kube-apiserver is vulnerable to memory explosion.
+The issue is apparent in larger clusters, where only a few LIST requests might cause serious disruption.
+Uncontrolled and unbounded memory consumption of the servers does not only affect clusters that operate in an 
+HA mode but also other programs that share the same machine.
+In this KEP we propose a potential solution to this issue.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today informers are the primary source of LIST requests.
+The LIST is used to get a consistent snapshot of data to build up a client-side in-memory cache.
+The primary issue with LIST requests is unpredictable memory consumption.
+The actual usage depends on many factors like the page size, applied filters (e.g. label selectors), query parameters, and sizes of individual objects.
+See the [Appendix](#appendix) section for more details on potential sources of LIST request and their impact on memory.
+![memory consumption during the synthetic test](./memory_usage_before.png "memory consumption during the synthetic test")
+In extreme cases, the server can allocate hundreds of megabytes per request.
+To better visualize the issue let's consider the above graph.
+It shows the memory usage of an API server during a test (see [manual test](#manual-testing-without-the-changes-in-place) section for more details).
+We can see that increasing the number of informers drastically increases the memory consumption of the server.
+Moreover, around 16:40 we lost the server after running 16 informers. During an investigation, we realized that the server allocates a lot of memory for handling LIST requests.
+In short, it needs to bring data from the database, unmarshal it, do some conversions and prepare the final response for the client.
+The bottom line is around O(5*the_response_from_etcd) of temporary memory consumption.
+Neither priority and fairness nor Golang garbage collection is able to protect the system from exhausting memory.
```

Comment:
APF should be able to limit a collection to 1 concurrent LIST (i.e., serialize all list calls). You're saying there's a condition where apiserver cannot handle a single LIST call?